# Data Joining

```{block, type="tmbinfo", echo=T}
> Creating an object that joins the metadata, microbial abundance, and taxonomy data simplifies the downstream analyses. The composite objects are in long-format and enables us to do any type of analysis we desire when exploring the results associated with **16S rRNA gene** or **metagenomics** sequencing data.
```

## Prerequisite
- Tidy metadata.
- Tidy microbial abundance table.
- Tidy taxonomy table.

## Load libraries and data
```{r message=FALSE, warning=FALSE}
source("R/common.R")
library(tidyverse)
library(glue)

load("../iMAP-part1/RDataRDS/metadata_objects.RData", verbose = T)
load("RDataRDS/phylo_otu_taxonomy_objects.RData", verbose = T)
# load("RDataRDS/opti_otu_taxonomy_objects.RData", verbose = T)
load("RDataRDS/q2_otu_taxonomy_objects.RData", verbose = T)
load("RDataRDS/m3_otu_taxonomy_objects.RData", verbose = T)
```

## Creating R objects from `mothur` output {#mothur-composite}
```{r}
taxlevels <- c("kingdom", "phylum", "class", "order", "family", "genus" )

metadata <-  mo_metadata %>% 
  select(sample_id, isolate) %>% 
  drop_na(isolate)

paste("Dimension of the metadata table:", 
      dim(metadata)[1], "rows and", 
      dim(metadata)[2], "columns")


otu_counts <- otutable %>% 
  pivot_longer(-otu, names_to = "sample_id", values_to = "count") %>% 
  relocate(sample_id, .before = otu)

paste("Dimension of the otutable table:", 
      dim(otu_counts)[1], "rows and", 
      dim(otu_counts)[2], "columns")


taxonomy <- taxonomy %>% 
  mutate(pretty_otu = str_replace(string = otu,
                         pattern = "phylo0*",
                         replacement = "Phylotyp ")) %>% 
  relocate(pretty_otu, .after = "otu")


paste("Dimension of the taxonomy table:", 
      dim(taxonomy)[1], "rows and", 
      dim(taxonomy)[2], "columns")

composite <- inner_join(metadata, otu_counts, by = "sample_id") %>% 
  inner_join( ., taxonomy, by = "otu") %>% 
  group_by(sample_id) %>% 
  mutate(rel_abund = count/sum(count)) %>% 
  ungroup() %>% 
  relocate(count, .before = rel_abund) 

colnames(composite)

## Confirm that rel abund adds to 1 (100%)
composite %>% 
  group_by(sample_id) %>% 
  summarise(total_re_abund = sum(rel_abund))

## Create tidy R object
Rjoined_object <- composite %>% 
  pivot_longer(c(all_of(taxlevels), otu), names_to = "level", values_to = "taxon") %>% 
  relocate(rel_abund, .after = taxon) %>% 
  mutate(taxon = str_replace(string = taxon,
                            pattern = "(.*)",
                            replacement = "*\\1*"),
        taxon = str_replace(string = taxon,
                            pattern = "\\*(.*)_unclassified\\*",
                            replacement = "Unclassified<br>*\\1*"),
        taxon = str_replace_all(taxon, "_", " "),
        taxon_phylo = glue("{taxon}<br>({pretty_otu})"))


saveRDS(composite, "RDataRDS/composite.rds")
save(composite, Rjoined_object, file = "RDataRDS/Rjoined_objects.RData")
save(composite, Rjoined_object, file = "RDataRDS/saved_objects.RData")
```

## Preparing data for statistical analysis {#stats-data}
```{r subsample}
metadata <- readRDS("~/Dropbox/CDILLC/GIT_REPOS/smda-end2end/RDataRDS/mo_metadata.rds") %>% 
  select(sample_id, group = isolate) %>% 
  drop_na(group)

phylo_sub_counts_long <- read_tsv("~/Dropbox/CDILLC/GIT_REPOS/smda-end2end/data/final.tx.1.subsample.shared", show_col_types = F) %>%
  rename_all(tolower) %>% 
  select(-label, -numotus, sample_id = group, starts_with("phylo")) %>%
  pivot_longer(-sample_id, names_to="otu", values_to="count") %>% 
  filter(count != 0)
saveRDS(phylo_sub_counts_long, "RDataRDS/phylo_sub_counts_long.rds")

otutable <- phylo_sub_counts_long %>%
  pivot_wider(id_cols = otu, names_from = sample_id, values_from = count) %>%
  mutate_if(is.numeric, ~replace(., is.na(.), 0))

paste("Dimension of the phylotype subsample table table is", 
      dim(otutable)[1], "rows and", 
      dim(otutable)[2], "columns") 
saveRDS(otutable, "RDataRDS/phylo_sub_table.rds")


nseqs_per_sample <- phylo_sub_counts_long %>% 
  group_by(sample_id) %>% 
  summarise(N = sum(count), .groups = "drop") %>% 
  count(N) %>% 
  pull(N)

stopifnot(length(nseqs_per_sample) == 1)

lod <- 100*1/nseqs_per_sample


taxlevels <- c("kingdom", "phylum", "class", "order", "family", "genus")
taxonomy <- read_tsv("~/Dropbox/CDILLC/GIT_REPOS/smda-end2end/data/final.tx.1.cons.taxonomy", show_col_types = F) %>%
  rename_all(tolower) %>%
  select(otu, taxonomy) %>%
  mutate(otu = tolower(otu),
         taxonomy = str_replace_all(taxonomy, "\\(\\d+\\)", ""),
         taxonomy = str_replace_all(taxonomy, ";$", "")) %>%
  mutate(pretty_otu = str_replace(string = otu,
                         pattern = "phylo0*",
                         replacement = "Phylotype ")) %>%
  relocate(pretty_otu, .after = "otu") %>%
  separate(taxonomy, into = all_of(taxlevels), sep = ";")

paste("Dimension of the taxonomy table is",
      dim(taxonomy)[1], "rows and",
      dim(taxonomy)[2], "columns")

head(taxonomy)[, 1:4] %>% as.data.frame()
saveRDS(taxonomy, "RDataRDS/sub_phylo_taxonomy.rds")


sub_composite <- inner_join(metadata, phylo_sub_counts_long, by = "sample_id") %>% 
  inner_join(., taxonomy, by = "otu") %>% 
  group_by(sample_id) %>% 
  mutate(rel_abund = count/sum(count)) %>% 
  ungroup() %>% 
  relocate(count, .before = rel_abund) 

colnames(sub_composite)

## Confirm that rel abund adds to 1 (100%)
sub_composite %>% 
  group_by(sample_id) %>% 
  summarise(total_re_abund = sum(rel_abund))

## Create tidy R object
Rjoined_sub_object <- sub_composite %>% 
  pivot_longer(c(all_of(taxlevels)), names_to = "level", values_to = "taxon") %>% 
  relocate(rel_abund, .after = taxon) %>% 
  mutate(taxon = str_replace(string = taxon,
                            pattern = "(.*)",
                            replacement = "*\\1*"),
        taxon = str_replace(string = taxon,
                            pattern = "\\*(.*)_unclassified\\*",
                            replacement = "Unclassified<br>*\\1*"),
        taxon = str_replace_all(taxon, "_", " "),
        taxon_phylo = glue("{taxon}<br>({pretty_otu})"))


saveRDS(sub_composite, "RDataRDS/sub_composite.rds")
resave(sub_composite, Rjoined_sub_object, file = "RDataRDS/Rjoined_objects.RData")
resave(sub_composite, Rjoined_sub_object, file = "RDataRDS/saved_objects.RData")
```

## Creating R objects from `qiime2` output {#qiime2-composite}
```{r}
taxlevels <- c("kingdom", "phylum", "class", "order", "family", "genus" )

metadata <- q2_metadata %>% 
  select(sample_id = "sample-id", isolate) %>% 
  drop_na(isolate)

otu_counts <- q2_otutable %>% 
  pivot_longer(-feature, names_to = "sample_id", values_to = "count") %>% 
  relocate(sample_id, .before = feature)

taxonomy <- q2_taxonomy 
  
q2_composite <- inner_join(metadata, otu_counts, by = "sample_id") %>% 
  inner_join( ., taxonomy, by = "feature") %>% 
  group_by(sample_id) %>% 
  mutate(rel_abund = count/sum(count)) %>% 
  ungroup() %>% 
  relocate(count, .before = rel_abund)

cat("\nColumn names of the composite file\n")
colnames(q2_composite)

cat("\nDoes the relative abundance adds to 1? Must!\n")
q2_composite %>% 
  group_by(sample_id) %>% 
  summarise(total_re_abund = sum(rel_abund))

q2_Rjoined_object <- q2_composite %>% 
  pivot_longer(cols = c(all_of(taxlevels), "feature"), names_to = "level", values_to = "taxon") %>% 
  relocate(c(count, rel_abund), .after = taxon)

saveRDS(q2_composite, "RDataRDS/q2_composite.rds")
resave(q2_Rjoined_object, file = "RDataRDS/Rjoined_objects.RData")
resave(q2_composite, q2_Rjoined_object, file = "RDataRDS/saved_objects.RData")
```


## Creating R objects from `metaphlan` output  {#metaphlan-composite}
- Here we create a composite object using default relative abundance output.
```{r}
taxlevels <- c("kingdom", "phylum", "class", "order", "family", "genus" )

metadata <- m3_metadata %>% 
  select(sample_id, platform) %>% 
  drop_na(platform)

m3_rel_otutable <- m3_otutable %>% 
  pivot_longer(-otu, names_to = "sample_id", values_to = "rel_abund") %>% 
  relocate(sample_id, .before = otu)

m3_glued_taxonomy <- m3_taxonomy %>% 
  mutate(pretty_otu = str_replace(string = otu,
                             pattern = "phylo0*",
                             replacement = "Phylo "),
        genus = str_replace(string = genus,
                            pattern = "(.*)",
                            replacement = "*\\1*"),
        genus = str_replace(string = genus,
                            pattern = "\\*(.*)_unclassified\\*",
                            replacement = "Unclassified<br>*\\1*"),
        genus_phylo = glue("{genus}<br>({pretty_otu})"))

m3_rel_composite <- inner_join(metadata, m3_rel_otutable, by = "sample_id") %>% 
  inner_join( ., m3_glued_taxonomy, by = "otu")

cat("\nColumn names of the composite file\n")
colnames(m3_rel_composite)

cat("\nDoes the relative abundance adds to 1? Must!\n")
m3_rel_composite %>% 
  group_by(sample_id) %>% 
  summarise(total_re_abund = round(sum(rel_abund)))

m3_rel_Rjoined_object <- m3_rel_composite %>% 
  pivot_longer(cols = c(all_of(taxlevels), "otu"), names_to = "level", values_to = "taxon") %>% 
  relocate(rel_abund, .after = taxon)


saveRDS(m3_rel_composite, "RDataRDS/m3_rel_composite.rds")
resave(m3_rel_Rjoined_object, file = "RDataRDS/Rjoined_objects.RData")
resave(m3_rel_composite, m3_rel_Rjoined_object, file = "RDataRDS/saved_objects.RData")
```

## Creating R objects from `metaphlan` estimated read counts {#metaphlan-count-composite}
- Here we create a composite object containing abundance based on the estimated mapped reads.

```{r}
taxlevels <- c("kingdom", "phylum", "class", "order", "family", "genus" )

metadata <- m3_metadata %>% 
  select(sample_id, platform) %>% 
  drop_na(platform)

m3_read_based_otutable <-  m3_read_counts_otutable %>% 
  pivot_longer(-otu, names_to = "sample_id", values_to = "count") %>% 
  relocate(sample_id, .before = otu) %>% 
  group_by(sample_id) %>% 
  mutate(rel_abund = count/sum(count)) %>% 
  ungroup() %>% 
  relocate(count, .before = rel_abund) 

m3_count_glued_taxonomy <- m3_taxonomy %>% 
  mutate(pretty_otu = str_replace(string = otu,
                             pattern = "otu0*",
                             replacement = "OTU "),
        genus = str_replace(string = genus,
                            pattern = "(.*)",
                            replacement = "*\\1*"),
        genus = str_replace(string = genus,
                            pattern = "(.*)_unclassified*",
                            replacement = "Unclassified<br>*\\1"),
        genus_phylo = glue("{genus}<br>({pretty_otu})"))

m3_read_count_composite <- inner_join(metadata, m3_read_based_otutable, by = "sample_id") %>% 
  inner_join( ., m3_count_glued_taxonomy, by = "otu")

cat("\nColumn names of the composite file\n")
colnames(m3_read_count_composite)

cat("\nDoes the relative abundance adds to 1? Must!\n")
m3_read_count_composite %>% 
  group_by(sample_id) %>% 
  summarise(total_re_abund = round(sum(rel_abund)))

m3_read_count_Rjoined_object <- m3_read_count_composite %>% 
  pivot_longer(cols = c(all_of(taxlevels), "otu"), names_to = "level", values_to = "taxon") %>% 
  relocate(rel_abund, .after = taxon)

saveRDS(m3_read_count_composite, "RDataRDS/m3_read_count_composite.rds")
resave(m3_read_count_Rjoined_object, file = "RDataRDS/Rjoined_objects.RData")
resave(m3_read_count_Rjoined_object, file = "RDataRDS/saved_objects.RData")
```

## Review saved R joined objects
```{r}
load("RDataRDS/Rjoined_objects.RData", verbose = T)
```

# Pyloseq objects {#phyloseq-objects}

## Prerequisite
- Tidy metadata created in [PART 1](https://github.com/tmbuza/microbiome-part1/).
- Tidy microbial abundance table created in [PART 2](https://github.com/tmbuza/microbiome-part2/).
- Tidy taxonomy table created in [PART 2](https://github.com/tmbuza/microbiome-part2/).
- Synthetic tidy data

## Load libraries and data
```{r message=FALSE, warning=FALSE}
source("R/common.R")
library(tidyverse)

cat("Metadata\n\n")
load("../iMAP-part1/RDataRDS/metadata_objects.RData", verbose = T)

cat("\nOtu and taxonomy tables\n\n")
load("RDataRDS/phylo_otu_taxonomy_objects.RData", verbose = T)
load("RDataRDS/q2_otu_taxonomy_objects.RData", verbose = T)
load("RDataRDS/m3_otu_taxonomy_objects.RData", verbose = T)
```

## Using `mothur` output {#mothur-physeq}
```{r}
ps_metadata <- metadata %>% tibble::column_to_rownames("sample_id") %>% 
  sample_data(ps_metadata)

ps_otutable <- otutable %>% tibble::column_to_rownames("otu") %>% 
  otu_table(ps_otutable, taxa_are_rows = TRUE)

ps_taxonomy <- as.matrix(taxonomy %>% tibble::column_to_rownames("otu")) %>% 
  tax_table(ps_taxonomy)

ps_raw <- merge_phyloseq(ps_metadata, ps_otutable, ps_taxonomy)

ps_taxa_tree = rtree(ntaxa(ps_raw), rooted=TRUE, tip.label=taxa_names(ps_raw))

## Optionally view the tree before merging to ps object
# plot(ps_taxa_tree) 

ps_raw <- phyloseq::merge_phyloseq(ps_raw, ps_taxa_tree)
ps_rel <- phyloseq::transform_sample_counts(ps_raw, function(x){x / sum(x)})

ps_raw
save(ps_raw, ps_rel,  file = "RDataRDS/phyloseq_objects.RData")
```

Summarizing `mothur` phyloseq object
```{r message=FALSE, warning=FALSE}
microbiome::summarize_phyloseq(ps_raw)
```

## Using `qiime2` output {#qiime2-physeq}
```{r}
ps_q2_metadata <- q2_metadata %>% 
  rename(sample_id = "sample-id") %>% 
  tibble::column_to_rownames("sample_id") %>% 
  sample_data(ps_q2_metadata)

ps_q2_otutable <- q2_otutable %>% tibble::column_to_rownames("feature") %>% 
  otu_table(ps_q2_otutable, taxa_are_rows = TRUE)

ps_q2_taxonomy <- as.matrix(q2_taxonomy %>% tibble::column_to_rownames("feature")) %>% 
  tax_table(ps_q2_taxonomy)

ps_q2 <- merge_phyloseq(ps_q2_metadata, ps_q2_otutable, ps_q2_taxonomy)

ps_q2_taxa_tree = rtree(ntaxa(ps_q2), rooted=TRUE, tip.label=taxa_names(ps_q2))
## Optionally view the tree before merging to ps object
# plot(ps_q2_taxa_tree) 

ps_q2_raw <- phyloseq::merge_phyloseq(ps_q2, ps_q2_taxa_tree)
ps_q2_rel <- phyloseq::transform_sample_counts(ps_q2_raw, function(x){x / sum(x)})

ps_q2_raw
resave(ps_q2_raw, ps_q2_rel,  file = "RDataRDS/phyloseq_objects.RData")
```

Summarizing `qiime2` phyloseq object
```{r message=FALSE, warning=FALSE}
microbiome::summarize_phyloseq(ps_q2_raw)
```

## Using `metaphlan` output {#metaphlan-physeq}
```{r}
ps_m3_metadata <- m3_metadata %>% tibble::column_to_rownames("sample_id") %>% 
  sample_data(ps_m3_metadata)

ps_m3_otutable <- m3_read_counts_otutable %>% tibble::column_to_rownames("otu") %>% 
  otu_table(ps_m3_otutable, taxa_are_rows = TRUE)

ps_m3_taxonomy <- as.matrix(m3_read_counts_taxonomy %>% tibble::column_to_rownames("otu")) %>% 
  tax_table(ps_m3_taxonomy)

ps_m3 <- merge_phyloseq(ps_m3_metadata, ps_m3_otutable, ps_m3_taxonomy)

ps_m3_taxa_tree = rtree(ntaxa(ps_m3), rooted=TRUE, tip.label=taxa_names(ps_m3))
# plot(ps_m3_taxa_tree)

ps_m3_raw <- phyloseq::merge_phyloseq(ps_m3, ps_m3_taxa_tree)
ps_m3_rel <- phyloseq::transform_sample_counts(ps_m3_raw, function(x){x / sum(x)})

ps_m3_raw
resave(ps_m3_raw, ps_m3_rel,  file = "RDataRDS/phyloseq_objects.RData")
```

## Synthetic phyloseq data
```{r}
library("phyloseq")
library("tidyverse")
library("glue")
set.seed(2022)

# Toy otu table
toy_otu_mtx <- matrix(sample(1:1000, 100, replace = TRUE), nrow = 10, ncol = 10)
rownames(toy_otu_mtx) <- paste0("OTU", 1:nrow(toy_otu_mtx))
colnames(toy_otu_mtx) <- paste0("Sample", 1:ncol(toy_otu_mtx))

# Toy taxonomy table
toy_tax_mtx <- matrix(sample(letters, 70, replace = TRUE), nrow = nrow(toy_otu_mtx), ncol = 7)
rownames(toy_tax_mtx) <- rownames(toy_otu_mtx)
colnames(toy_tax_mtx) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

# Basic phyloseq object
OTU <- otu_table(toy_otu_mtx, taxa_are_rows <- TRUE)
TAX <- tax_table(toy_tax_mtx)
ps_toy <- phyloseq(OTU, TAX)

# Toy metadata
toy_metadata <- sample_data(
  data.frame(Var1 = sample(LETTERS[1:2],
                               size=nsamples(ps_toy), 
                               replace=TRUE),
             Var2 = sample(LETTERS[5:7],
                               size=nsamples(ps_toy), 
                               replace=TRUE),
             row.names = sample_names(ps_toy),
             stringsAsFactors = FALSE))
toy_metadata$Var2 <- paste(toy_metadata$Var1, toy_metadata$Var2, sep="_")

ps_toy <- merge_phyloseq(toy_metadata, ps_toy)

# Toy tree
library("ape")
ps_toy_taxa_tree = rtree(ntaxa(ps_toy), rooted=TRUE, tip.label=taxa_names(ps_toy))
# plot(ps_toy_taxa_tree)

ps_toy_raw <- phyloseq::merge_phyloseq(ps_toy, ps_toy_taxa_tree)
ps_toy_rel <- phyloseq::transform_sample_counts(ps_toy_raw, function(x){x / sum(x)})

ps_toy_raw
resave(ps_toy_raw, ps_toy_rel,  file = "RDataRDS/phyloseq_objects.RData")
```



## Review saved phyloseq objects
```{r}
load("RDataRDS/phyloseq_objects.RData", verbose = T)
```


```{r eval=FALSE, fig.height=10, fig.width=10, include=FALSE}
## Quick glimpse on ps tree <-- See microbiome-part4, 5d...
load("RDataRDS/phyloseq_objects.RData")

plot_tree(ps_raw, 
          color="isolate", 
          label.tips="taxa_names", 
          ladderize="left", 
          plot.margin=0.3)
plot_heatmap(ps_raw)
plot_tree(ps_q2_raw, 
          color="isolate", 
          label.tips="taxa_names", 
          ladderize="left", 
          plot.margin=0.3)
plot_heatmap(ps_q2_raw)
plot_tree(ps_m3_raw, 
          color="platform", 
          shape="direction", 
          label.tips="taxa_names", 
          ladderize="left", 
          plot.margin=0.3)
plot_heatmap(ps_m3_raw)
plot_tree(ps_toy_raw, 
          color="Var2", 
          shape="Var1", 
          label.tips="taxa_names", 
          ladderize="left", 
          plot.margin=0.3)
plot_heatmap(ps_toy_raw)
```


# Hypertension Data 
- The code below is in R/genus_process.R
- Use source("R/genus_process.R") to load it

```{r}
library(tidyverse)
library(broom)
library(ggtext)
library(data.table)

set.seed(2022)

shared <- read_csv("../Datasets/HypertensionProject.csv", show_col_types = F) %>%
  dplyr::select(1, Prevotella:ncol(.)) %>%
  transpose(keep.names = "taxonomy", make.names = "SampleID") %>%
  # select(taxonomy, starts_with("ERR")) %>%
  pivot_longer(-taxonomy, names_to="sample_id", values_to="rel_abund") %>%
  relocate(sample_id)

metabolites <- read_csv("../Datasets/HypertensionProject.csv", show_col_types = F) %>%
  dplyr::select(c(1,5:18 )) %>%
  transpose(keep.names = "metabopwy", make.names = "SampleID") %>%
  # select(metabopwy, starts_with("ERR")) %>%
  pivot_longer(-metabopwy, names_to="sample_id", values_to="value") %>%
  relocate(sample_id)

taxonomy <- read_tsv("../Datasets/mo_demodata/baxter.cons.taxonomy", show_col_types = F) %>%
  rename_all(tolower) %>%
  select(otu, taxonomy) %>%
  mutate(otu = tolower(otu),
         taxonomy = str_replace_all(taxonomy, "\\(\\d+\\)", ""),
         taxonomy = str_replace(taxonomy, ";unclassified", "_unclassified"),
         taxonomy = str_replace_all(taxonomy, ";unclassified", ""),
         taxonomy = str_replace_all(taxonomy, ";$", ""),
         taxonomy = str_replace_all(taxonomy, ".*;", "")
  )

metadata <- read_csv("../Datasets/HypertensionProject.csv", show_col_types = F) %>%
  dplyr::select(c(1:3)) %>%
  mutate(dsestate = Disease_State == "HTN" | Disease_State == "pHTN",
         control = Disease_State == "Control") %>%
  rename(sample_id = SampleID)

## Data joining

composite <- inner_join(shared, metadata, by="sample_id")

metabo_composite <- inner_join(shared, metabolites, by="sample_id") %>%
  group_by(sample_id, metabopwy) %>%
  summarise(value = sum(value), .groups="drop") %>%
  group_by(sample_id) %>%
  mutate(metabo_rel_abund = value / sum(value)) %>%
  ungroup() %>%
  select(-value) %>%
  inner_join(., metadata, by="sample_id")
```



